scripts/generate_rust_analyzer.py

```diff
@@ -0,0 +1,175 @@
+    for x in list:
+        pos += 1
+        if x['display_name'] == name:
+            return pos
+
+def generate_deps(source, deps):
+    ret = []
+    for x in deps:
+        ret.append({"crate": find(source, x), "name": x})
+    return ret
+
+def generate_cfgs():
+    ret = []
+    with open( objtree + "/include/generated/rustc_cfg") as f:
+        for line in f:
+            line = line.replace('--cfg=', '')
+            line = line.replace('\n', '')
+            ret.append(line)
+    return ret
+
+def get_default_deps(source = crate_list):
+    return generate_deps(source, ["core", "alloc", "kernel"])
+
+def generate_crate(name, module, member = True, deps = get_default_deps(), cfg = generate_cfgs(), extra = {}, edition = "2018"):
+    return {
+        "display_name": name,
+        "root_module": module,
+        "edition": edition,
+        "is_workspace_member": member,
+        "cfg": cfg,
+        "deps": deps,
+        "env": {
+            "RUST_MODFILE": "This is only for rust-analyzer"
+        }
+    }
+
+def generate_kernel(srctree = srctree, objtree = objtree, lib_src = lib_src, bindgen_file = bindgen_file):
+    ret = []
+
+    # core
+    ret.append(
+        generate_crate(
+            name="core",
+            module=lib_src + "/core/src/lib.rs",
+            member=False,
+            deps=[],
+            cfg=[]
+        )
+    )
+
+    # compiler_builtins
+    ret.append(
+        generate_crate(
+            name="compiler_builtins",
+            module=srctree + "/rust/compiler_builtins.rs",
+            deps=[],
+            cfg=[]
+        )
+    )
+
+    # alloc
+    ret.append(
+        generate_crate(
+            name="alloc",
+            module=lib_src + "/alloc/src/lib.rs",
+            member=False,
+            deps=generate_deps(ret, ["core", "compiler_builtins"]),
+            cfg=[]
+        )
+    )
+
+    # module
+    module = generate_crate(
+        name="module",
+        module=srctree + "/rust/module.rs",
+        deps=[],
+        cfg=[],
+    )
+    module["proc_macro_dylib_path"] = "rust/libmodule.so"
+    ret.append(
+        module
+    )
+
+    # kernel
+    kernel = generate_crate(
+        name="kernel",
+        module=srctree + "/rust/kernel/lib.rs",
+        deps=generate_deps(ret, ["core", "alloc", "module"]),
+    )
+    kernel["env"]["RUST_BINDINGS_FILE"] = bindgen_file
+    kernel["source"] = {
+        "include_dirs": [ srctree + "/rust/kernel", objtree + "/rust" ],
+        "exclude_dirs": []
+    }
+    ret.append(kernel)
+
+    return ret
+
+def does_file_contains(filepath, search):
+    with open (filepath) as f:
+        for line in f:
+            if search in line:
+                return True
+    return False
+
+def check_and_generate(filepath, source=crate_list):
+    file = path.basename(filepath)
+    makefile = path.dirname(filepath) + "/Makefile"
+    name = file.replace('.rs', '')
+    objname = name + ".o"
+    print("checking " + objname, file=sys.stderr)
```

Comment:
For templating, avoid string concatenation. Instead, there is the `%` operator (in the past, for simple cases), the `format()` method (for bigger/complex cases) and nowadays, f-strings (similar to `format()` but for the inline cases too, better than the `%` operator):

```python
a = 42
'hello %s there' % a # the % operator
'hello {a} there'.format(a=a) # the .format() method
f'hello {a} there'   # f-strings
```

But since this is for `stderr` logging, give a try to `logging` instead:

```python
logging.info("hello %s there", a)
```

For small scripts, I usually configure its output in `main()` as:

```python
logging_level = logging.DEBUG if verbose else logging.INFO
logging.basicConfig(format='[%(asctime)s] [%(levelname)s] %(message)s', level=logging_level)
```

Which gives this nice log:

```
[2021-05-05 14:38:17,325] [INFO] hello 42 there
```

Reply:
Isn't logging a bit much for this small (mostly) one shot script?

Reply:
When you write a lot of similar Python scripts, you realize it is easier to just always use `argparse`, `logging` and friends rather than "upgrade" them later, and so one ends up `cp`ing a `template.py` file with the handful lines of setup every time you write a script :-)

But yeah, without context, it can be seen as overkill.